### ISSUE TITLE
NuGet: Adding a .NET Standard assembly with RuntimeIdentifier=win will override other TFM assemblies (.NET, Android, iOS, UWP)

### ISSUE THREAD

**Issue by xen2** [NuGet/Home#7661]
TL;DR: Adding a runtime-specific assembly for .NET Standard overrides other TFM assemblies such as .NET, Android, UWP, iOS, etc...

I made a package with following files:

```
ref\netstandard2.0\Lib1.dll
runtimes\win\lib\netstandard2.0\Lib1.dll
ref\net45\Lib1.dll
lib\net45\Lib1.dll
```

If I compile a `net45` assembly using either no `RuntimeIdentifier` or `RuntimeIdentifier=win7` it will resolve those:

**(A)**
```
Output Item(s): 
    RuntimeCopyLocalItems=
        C:\Users\xen\.nuget\packages\testpackage\1.0.2\lib\net45\Lib1.dll
                NuGetPackageId=testpackage
                NuGetPackageVersion=1.0.2
Output Item(s): 
    ResolvedCompileFileDefinitions=
        C:\Users\xen\.nuget\packages\testpackage\1.0.2\ref\net45\Lib1.dll
                ExternallyResolved=true
                HintPath=C:\Users\xen\.nuget\packages\testpackage\1.0.2\ref\net45\Lib1.dll
                NuGetPackageId=testpackage
                NuGetPackageVersion=1.0.2
                NuGetSourceType=Package
                Private=false
```

However, as soon as I use `RuntimeIdentifier=win`, the `netstandard2.0/win` assembly takes predecence over `net45` one:

**(B)**
```
Output Item(s): 
    RuntimeCopyLocalItems=
        C:\Users\xen\.nuget\packages\testpackage\1.0.2\runtimes\win\lib\netstandard2.0\Lib1.dll
                NuGetPackageId=testpackage
                NuGetPackageVersion=1.0.2
Output Item(s): 
    ResolvedCompileFileDefinitions=
        C:\Users\xen\.nuget\packages\testpackage\1.0.2\ref\net45\Lib1.dll
                ExternallyResolved=true
                HintPath=C:\Users\xen\.nuget\packages\testpackage\1.0.2\ref\net45\Lib1.dll
                NuGetPackageId=testpackage
                NuGetPackageVersion=1.0.2
                NuGetSourceType=Package
                Private=false
```

It is not clear what is the expected priority order:
(1) net45/win, net45, netstandard2.0/win, netstandard2.0?
(2) net45/win, netstandard2.0/win, net45, netstandard2.0?

For me (1) makes more sense. (2) seems pretty bad to deal with additional TFM: adding a `netstandard2.0/win` would override everything since UWP uses `win` and MonoAndroid/iOS either don't have a runtime identifier or use `win`).

Either way, in both cases (1) and (2), it seems there is something not working as expected:

* Case (1): all situations should end up with (A) (right now it doesn't when `RuntimeIdentifier=win`)
* Case (2): it seems something is wrong with RID inheritance being ignored (`RuntimeIdentifier=win7` should end up with (B) rather than (A)).

On top of that I ran into some other discrepencies between `ResolvePackageAssets` and `ResolveNuGetPackageAssets` (esp. when no `RuntimeIdentifier` is set, it ends up finding only ref assemblies without looking into lib folder). I will detail those further once this first part is cleared up.

**Comment by xen2** [comment]
Quick addition: https://github.com/NuGet/NuGet.BuildTasks/blob/visual-studio-2015-update-3/src/Microsoft.NuGet.Build.Tasks/ResolveNuGetPackageAssets.cs#L343 assumes that if no runtime identifier is given, then there is nothing to copy.
This doesn't work well on platform such as Android which don't define a `NuGetRuntimeIdentifier`.

**Comment by xen2** [comment]
After some investigating, it looks like `project.assets.json` is generated with priority order (2) net45/win, netstandard2.0/win, net45, netstandard2.0.

Is this really intended? This makes any multi-TFM assembly which uses netstandard2.0/win unable to use any other TFM that depends on `win` due to netstandard2.0/win overriding other TFM (UWP uses `win10` which inherit from `win`, and Android/iOS fallback to `win`).

This happens in `LockFileUtils.GetLockFileItems()` -> `ContentItemCollection.FindBestItemGroup()`, because those TFM have .NET Standard as a fallback so first criteria `.NET4.5/win` also matches with `.NET Standard/win` without going to next criteria `.NET Standard (no RID)`.

Maybe the fallback shouldn't be automatic and criteria should be explicitely expanded/ordered?

**Comment by nkolev92** [comment]
Does this answer your question? 

https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders

**Comment by xen2** [comment]
Thanks, it matches the behavior I saw.
So what's the intended usage if I want to make a `.NET Standard/win` + `.NET Standard/linux` + `MonoAndroid` + `UAP` + `iOS` nupkg?

Note that amongst those:
* `iOS` define `win`
* `UAP` defines `win10`
* `MonoAndroid` defines nothing but copy nothing due to https://github.com/NuGet/NuGet.BuildTasks/blob/visual-studio-2015-update-3/src/Microsoft.NuGet.Build.Tasks/ResolveNuGetPackageAssets.cs#L343 resulting in ref being copied/used at runtime (note: I currently have to override that task with my own implementation until this is fixed)

I see two options for this `.NET Standard/win` + `.NET Standard/linux` + `MonoAndroid` + `UAP` + `iOS` nupkg:

- Pack `MonoAndroid` and `UAP` and `iOS` within `runtimes\win\` too. It makes sense for `UAP` but feels counter-intuitive for two others.
Is that really what I should do? (it feels quite strange to have to put Xamarin ones in `win` runtime).
Also worried if at some point either the NuGet behavior changes or if Xamarin removes this default `RuntimeIdentifier`.
On top of that, Android doesn't even have a default identifier defined.
- Move `.NET Standard/win` to `.NET Standard/win7` to avoid conflict with `MonoAndroid` & `iOS` default (they only define `RuntimeIdentifier=win` not `win7`).
This would however force us to add `<RuntimeIdentifier>win7</RuntimeIdentifier>` to old-style .NET 4.6.1 projects (it seems new csproj default to `win7` but old-style csproj default to `win`) but we could live with that.

Both options seem not so straightforward, so I was wondering, what would you recommend?

**Comment by indigozero** [comment]
Take a look at the following link if you haven't already. It kind of explains why scenario B happens as RID qualified assets are preferred. I'm thinking that in scenario A not having a RID would make it fall back on the TFM, but that doesn't explain the RID=win7 part of the scenario. Hope that helps?

https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/readme.md

**Comment by xen2** [comment]
Update: I went for option 1 and then found out I could use the runtime `any` (I was afraid if I chose `any` + exact TFM it would be overriden by `win` with a fallback TFM but it doesn't seem to be the case. The override is just about `runtimes` being present means `lib` is ignored).

Here is my final hierarchy:

```
ref\netstandard2.0\Lib1.dll
ref\monoandroid81\Lib1.dll
ref\uap10.0.16299\Lib1.dll
runtimes\win\lib\netstandard2.0\Lib1.dll
runtimes\linux\lib\netstandard2.0\Lib1.dll
runtimes\any\lib\monoandroid81\Lib1.dll
runtimes\any\lib\uap10.0.16299\Lib1.dll
```

The only issue left is that I need to add this for Android to work:
```
  <!-- If not set, Android executable doesn't work properly because ResolveNuGetPackageAssets has no NuGetRuntimeIdentifier and use only the ref assemblies -->
  <PropertyGroup Condition="'$(XenkoPlatform)' == 'Android' And '$(AndroidApplication)' == 'true'">
    <RuntimeIdentifier>win</RuntimeIdentifier>
    <NuGetRuntimeIdentifier>win</NuGetRuntimeIdentifier>
  </PropertyGroup>
```

**Comment by jainaashish** [comment]
I wasn't able to fully follow what you're trying to do.. But some of the suggestions and guidance:

- if you only have a single compile time implementation for your assembly then you can just put it under `ref/any/Lib1.dll`. You don't need duplicate copies in different tfms if they are identitical. Also, `/ref` is only for compile time.
- `/runtimes` is to understand runtime assemblies which is where you'd have different implementation based on different platforms.Something like:
runtimes/win/lib/netstandard2.0/lib1.dll
runtimes/linux/lib/netstandard2.0/lib1.dll
runtimes/win10/lib/netstandard2.0/lib1.dll
runtimes/linux/lib/monoandroid81/lib1.dll
.........

But i'd suggest not to have `/any` folder under `/runtimes` for platforms, you can have it for tfms which is what you won't be sure of.

**Comment by xen2** [comment]
What I am trying to do a proper package with several TFM and multiple runtime identifier on Windows.
Seems like I could manage to do that as mentioned in my previous message.
I will close this and report my last android issue to the proper Github (Mono.Android).

Thanks.
